geopandas/tests/test_op_output_types.py

```diff
@@ -0,0 +1,276 @@
+import pandas as pd
+import pyproj
+import pytest
+
+from shapely.geometry import Point
+
+from geopandas import GeoDataFrame, GeoSeries
+
+crsgs_osgb = pyproj.CRS(27700)
+crs_wgs = pyproj.CRS(27700)
+# TODO need to check that behaviour works with non default geom col name
+
+
+class TestDataFrameMethodReturnTypes:
+    def setup_method(self):
+        N = 10
+
+        self.df = GeoDataFrame(
+            [
+                {
+                    "value1": x + y,
+                    "value2": x * y,
+                    "geometry_temp": Point(x, y),  # rename this col in tests
+                }
+                for x, y in zip(range(N), range(N))
+            ],
+            crs=crs_wgs,
+            geometry="geometry_temp",
+        )
+        # want geometry2 to be a GeoSeries not Series, test behaviour of non geom col
+        self.df["geometry2"] = self.df["geometry_temp"].set_crs(
+            crsgs_osgb, allow_override=True
+        )
+
+    @staticmethod
+    def _check_metadata_gdf(gdf, geo_col_name="geometry", crs=crs_wgs):
+
+        assert gdf._geometry_column_name == geo_col_name
+        assert gdf.geometry.name == geo_col_name
+        assert gdf.crs == crs
+
+    @staticmethod
+    def _check_metadata_gs(gs, name="geometry", crs=crs_wgs):
+        assert gs.name == name
+        assert gs.crs == crs
+
+    def _check_standard_df(self, results, geo_col_name, method=None):
+        for i in results:
+            print(i.columns if isinstance(i, pd.DataFrame) else i.name)
+
+        assert type(results[0]) is pd.DataFrame
+        assert type(results[1]) is GeoDataFrame

```

Comment:
This depends on the order of the results passed to `_check_standard_df` inside `test_getitem`/`test_loc` etc, right? 
That might make it more difficult to follow what is exactly tested in those tests. 

Alternatively, we could maybe have a helper function for asserting an object is a DataFrame vs GeoDataFrame?  And then something like

```
self._assert_object(df[["value1", "value2"]], pd.DataFrame)
self._assert_object(df[[set_geom_col_name, "geometry2"]], GeoDataFrame, geo_column_name)
...
```

or 

```
# all cases that result in DataFrame
for result in [df[["value1", "value2"]], ...]:
    # ... call helper to assert result is a DataFrame

# all cases that result in a GeoDataFrame
for result in [df[[set_geom_col_name, "geometry2"]], ..]:
    # ... call helper to assert result is proper GeoDataFrame
```

I think something like that might make it easier to follow in the test itself, as it makes things more explicit. 
Of course, on the other hand, that gives a bit more duplication between `test_getitem`, `test_loc`, `test_iloc`, etc

Reply:
Yeah I did wonder about that point as I did this, guess I wanted to make clear that the same things are tested for `getitem`, `loc`, ... I think the duplication is okay if it makes thing easier to read / easier to examine when tests fail - coming from the perspective of trying to contribute code and work out why something is broken.

Reply:
Alternatively, we could also combine the getitem/loc/iloc cases in a single test. Like

```
for result in [
        df[[set_geom_col_name, "geometry2"]],
        df.loc[:, [set_geom_col_name, "geometry2"]],
        df.iloc[:, 2:4],
        ...
..]:
    # ... call helper to assert result is proper GeoDataFrame
```

Although I think it's fine to keep them separate. 

My main worry is that it's a bit complicated right now to read correctly what the `_check_standard_df` does for the different inputs (you need to check inside the function what it tests for the nth argument). So I would prefer making that more explicit (at the cost of a bit of duplication).